crates/ott-balancer/src/balancer.rs

```diff
@@ -171,12 +171,95 @@ impl BalancerLink {
     }
 }
 
-#[derive(Debug, Default)]
+#[derive(Debug)]
 pub struct BalancerContext {
     pub clients: HashMap<ClientId, BalancerClient>,
     pub monoliths: HashMap<MonolithId, BalancerMonolith>,
     pub rooms_to_monoliths: HashMap<RoomName, RoomLocator>,
     pub monoliths_by_region: HashMap<String, Vec<MonolithId>>,
+    pub monolith_selection: Box<dyn MonolithSelection + Send + Sync + 'static>,
+}
+#[derive(Debug, Default)]
+pub struct MonolithRegistry;
+pub trait MonolithSelection: std::fmt::Debug {
+    fn select_monolith(
+        &self,
+        monolith: Vec<&BalancerMonolith>,
+        ctx: &BalancerContext,
+    ) -> anyhow::Result<&BalancerMonolith>;
+
+    fn random_monolith(
+        &self,
+        monolith: Vec<&BalancerMonolith>,
+        ctx: &BalancerContext,
+    ) -> anyhow::Result<&BalancerMonolith>;
+}
+
+impl MonolithSelection for MonolithRegistry {
+    fn select_monolith(
+        &self,
+        _monolith: Vec<&BalancerMonolith>,
+        ctx: &BalancerContext,
+    ) -> anyhow::Result<&BalancerMonolith> {
+        fn cmp(x: &BalancerMonolith, y: &BalancerMonolith) -> std::cmp::Ordering {
+            x.rooms().len().cmp(&y.rooms().len())
+        }
+
+        let in_region = ctx
+            .monoliths_by_region
+            .get(BalancerConfig::get().region.as_str());
+        if let Some(in_region) = in_region {
+            let selected = in_region
+                .iter()
+                .flat_map(|id| ctx.monoliths.get(id))
+                .min_by(|x, y| cmp(x, y));
+            if let Some(s) = selected {
+                return Ok(s);
+            }
+        }
```

Comment:
... because this code should not be in the trait impl. The trait's purpose is not to do any filtering or preprocessing, its just to prioritize which monolith to use.